### PR TITLE
Add redfish emulator and containerize VBMC

### PR DESCRIPTION
**Pull request by maelk**
This is a first step towards supporting Redfish
- run VBMC as a container
- add a sushy-tools container to offer a redfish interface

What should be done next:
- add the docker images in quay.io repo and use those instead of building locally
- move the two containers to Kubernetes cluster (using secret for ssh key)
- modify the containers images, to not run as root etc.

### CODE REVIEW

**Comment by maelk** [comment]
related to issue https://github.com/metal3-io/metal3-dev-env/issues/47

File: resources/sushy-tools/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM python:3.7
+
+RUN apt update && apt install -y libvirt-dev && cd /root/ && \
+    git clone https://opendev.org/openstack/sushy-tools.git && \
+    cd /root/sushy-tools && \
+    pip3 install -r requirements.txt && python3 setup.py build && \
```
Comment (dtantsur) [inline resources/sushy-tools/Dockerfile]: Why not install `sushy-tools` via pip?

File: resources/vbmc/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,6 @@
+FROM python:3.7
+
+RUN apt update && \
+  apt install -y libvirt0 libvirt-dev && \
+  pip3 install virtualbmc
```
Comment (dtantsur) [inline resources/vbmc/Dockerfile]: nit: if you use centos as a base for this container, you can use python-virtualbmc from RDO.

File: resources/sushy-tools/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+    python3 setup.py install && pip3 install six requests libvirt-python
```
Comment (dtantsur) [inline resources/sushy-tools/Dockerfile]: Dependencies will be installed by pip, not need to do it explicitly.

Also, why not have a `CMD` here?

**Comment by dtantsur** [comment]
cc @etingof

File: resources/sushy-tools/Dockerfile (hunk 1 above)
Comment (maelk) [reply resources/sushy-tools/Dockerfile]: if we install sushy-tools via pip, the managers are not implemented (pip3 install sushy-tools==0.6.0). We need this feature, hence the installation from source

File: resources/sushy-tools/Dockerfile (hunk 3 above)
Comment (maelk) [reply resources/sushy-tools/Dockerfile]: six requests and libvirt-python are not in requirements.txt but are required to run sushy-emulator.
I'll add the CMD

File: resources/sushy-tools/Dockerfile (hunk 1 above)
Comment (dtantsur) [reply resources/sushy-tools/Dockerfile]: I'll request a release.

File: resources/sushy-tools/Dockerfile (hunk 3 above)
Comment (dtantsur) [reply resources/sushy-tools/Dockerfile]: I'll propose a fix for six and requests.

File: resources/sushy-tools/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.7
+
+RUN apt update && apt install -y libvirt-dev && cd /root/ && \
+    git clone https://opendev.org/openstack/sushy-tools.git && \
```
Comment (elfosardo) [inline resources/sushy-tools/Dockerfile]: you could also do
pip install git+https://opendev.org/openstack/sushy-tools.git

File: resources/sushy-tools/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.7
+
+RUN apt update && apt install -y libvirt-dev && cd /root/ && \
```
Comment (elfosardo) [inline resources/sushy-tools/Dockerfile]: nit: split one command per line

File: resources/vbmc/Dockerfile (hunk 2 above)
Comment (maelk) [reply resources/vbmc/Dockerfile]: does this bring something compared to pip installation ?

File: resources/sushy-tools/Dockerfile (hunk 1 above)
Comment (maelk) [reply resources/sushy-tools/Dockerfile]: Could you please wait for the release request ? we would need another fix in : https://review.opendev.org/#/c/677449/

File: resources/sushy-tools/Dockerfile (hunk 3 above)
Comment (maelk) [reply resources/sushy-tools/Dockerfile]: ok, I removed six and requests. we still need libvirt-python

File: resources/sushy-tools/Dockerfile (hunk 4 above)
Comment (maelk) [reply resources/sushy-tools/Dockerfile]: done, thank you

File: resources/sushy-tools/Dockerfile (hunk 5 above)
Comment (maelk) [reply resources/sushy-tools/Dockerfile]: done, thanks

File: resources/sushy-tools/Dockerfile (hunk 1 above)
Comment (dtantsur) [reply resources/sushy-tools/Dockerfile]: yeah, the release was delayed because of other outstanding issues

**Comment by maelk** [comment]
Any further comments ?

**Comment by russellb** [comment]
This will need to be rebased, sorry

**Comment by maelk** [comment]
@russellb rebased, thank you
@dtantsur thank you, now using pip3 install sushy-tools==0.7.0

**Comment by maelk** [comment]
We took a new image into use for CI. Re-running the tests

**Comment by maelk** [comment]
@russellb sorry for the CI mess on this one, we modified some things and the errors here are not related to the patchset

File: 01_install_requirements.sh [hunk 6]
```diff
@@ -21,13 +21,13 @@ if ! which docker-machine-driver-kvm2 2>/dev/null ; then
 fi
 
 if ! which kubectl 2>/dev/null ; then
-    curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl
+    curl -LO "https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl"
```
Comment (russellb) [inline 01_install_requirements.sh]: Can you split the unrelated quoting changes out into another patch?  That would make it easier to review the most important changes in this PR.

File: 01_install_requirements.sh [hunk 7]
```diff
@@ -59,3 +59,21 @@ ANSIBLE_FORCE_COLOR=true ansible-playbook \
   -e "virthost=$HOSTNAME" \
   -i vm-setup/inventory.ini \
   -b -vvv vm-setup/install-package-playbook.yml
+
+
+if [ ! -d "/etc/containers" ] ; then
```
Comment (russellb) [inline 01_install_requirements.sh]: What is the `/etc/containers` config needed for?  Can the Dockerfiles just be explicit about which registry to pull from?

File: centos_install_requirements.sh [hunk 8]
```diff
@@ -32,7 +32,7 @@ sudo yum -y install \
   wget
 
 # We're reusing some tripleo pieces for this setup so clone them here
-pushd $HOME
+pushd "$HOME"
```
Comment (russellb) [inline centos_install_requirements.sh]: It looks like all the changes in this file can just be removed, or split into another commit at least.

**Comment by russellb** [comment]
needs another rebase now, sorry about that.

**Comment by maelk** [comment]
metal3-dev-env is broken : https://github.com/metal3-io/metal3-dev-env/pull/59

**Comment by russellb** [comment]
I haven't tested the redfish part, but vbmc is still working for me :+1:

**Comment by hardys** [comment]
FYI I think this has broken dev-scripts, we need to start the new containers as in the 02 script here, although it's kind of unfortunate that we moved that out of the ansible role :(

Also we don't stop the virtualbmc.service on the host, so if you're testing in an existing environment everything looks OK.
